File: library/src/main/java/com/trinity/editor/VideoEditor.kt
/*
 * Copyright (C) 2019 Trinity. All rights reserved.
 * Copyright (C) 2019 Wang LianJie <wlanjie888@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.trinity.editor

import android.content.Context
import android.view.Surface
import android.view.SurfaceHolder
import android.view.SurfaceView
import com.tencent.mars.xlog.Log
import com.trinity.ErrorCode
import com.trinity.listener.OnRenderListener
import java.io.File

class VideoEditor(
  context: Context
) : TrinityVideoEditor, SurfaceHolder.Callback {

  companion object {
    const val NO_ACTION = -1
  }

  private var mId: Long = 0
  private var mFilterActionId = NO_ACTION
  // texture回调
  // 可以做特效处理 textureId是TEXTURE_2D类型
  private var mOnRenderListener: OnRenderListener ?= null

  init {
    val path = context.externalCacheDir?.absolutePath + "/resource.json"
    mId = create(path)
  }

  override fun setSurfaceView(surfaceView: SurfaceView) {
    surfaceView.holder.addCallback(this)
  }

  override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
    onSurfaceChanged(mId, width, height)
  }

  override fun surfaceDestroyed(holder: SurfaceHolder) {
    onSurfaceDestroyed(mId, holder.surface)
  }

  override fun surfaceCreated(holder: SurfaceHolder) {
    onSurfaceCreated(mId, holder.surface)
  }

  private external fun onSurfaceCreated(handle: Long, surface: Surface)

  private external fun onSurfaceChanged(handle: Long, width: Int, height: Int)

  private external fun onSurfaceDestroyed(handle: Long, surface: Surface)

  /**
   * 创建c++对象
   * @return 返回c++对象地址
   */
  private external fun create(resourcePath: String): Long

  /**
   * 获取视频时长
   */
  override fun getVideoDuration(): Long {
    return getVideoDuration(mId)
  }

  private external fun getVideoDuration(id: Long): Long

  override fun getCurrentPosition(): Long {
    return getCurrentPosition(mId)
  }

  private external fun getCurrentPosition(handle: Long): Long

  /**
   * 获取视频片段数量
   */
  override fun getClipsCount(): Int {
    return getClipsCount(mId)
  }

  private external fun getClipsCount(id: Long): Int

  /**
   * 获取视频clip
   * @param index 当前片段的下标
   * @return 下标无效时返回null
   */
  override fun getClip(index: Int): MediaClip? {
    return getClip(mId, index)
  }

  private external fun getClip(id: Long, index: Int): MediaClip?

  /**
   * 播入一段clip到队列
   * @param clip 插入的clip对象
   * @return 插入成功返回0
   */
  override fun insertClip(clip: MediaClip): Int {
    return insertClip(mId, clip)
  }

  private external fun insertClip(id: Long, clip: MediaClip): Int

  /**
   * 播入一段clip到队列
   * @param index 播入的下标
   * @param clip 插入的clip对象
   * @return 插入成功返回0
   */
  override fun insertClip(index: Int, clip: MediaClip): Int {
    return insertClip(mId, index, clip)
  }

  private external fun insertClip(id: Long, index: Int, clip: MediaClip): Int

  /**
   * 删除一个片段
   * @param index 删除片段的下标
   */
  override fun removeClip(index: Int) {
    removeClip(mId, index)
  }

  private external fun removeClip(id: Long, index: Int)

  /**
   * 替换一个片段
   * @param index 需要替换的下标
   * @param clip 需要替换片段的对象
   * @return 替换成功返回0
   */
  override fun replaceClip(index: Int, clip: MediaClip): Int {
    return replaceClip(mId, index, clip)
  }

  private external fun replaceClip(id: Long, index: Int, clip: MediaClip): Int

  /**
   * 获取所有片段
   * @return 返回当前所有片段
   */
  override fun getVideoClips(): List<MediaClip> {
    return getVideoClips(mId)
  }

  private external fun getVideoClips(id: Long): List<MediaClip>

  /**
   * 获取当前片段的时间段
   * @param index 获取的下标
   * @return 返回当前片段的时间段
   */
  override fun getClipTimeRange(index: Int): TimeRange {
    return getClipTimeRange(mId, index)
  }

  private external fun getClipTimeRange(id: Long, index: Int): TimeRange

  override fun getVideoTime(index: Int, clipTime: Long): Long {
    return getVideoTime(mId, index, clipTime)
  }

  private external fun getVideoTime(id: Long, index: Int, clipTime: Long): Long

  override fun getClipTime(index: Int, videoTime: Long): Long {
    return getClipTime(mId, index, videoTime)
  }

  private external fun getClipTime(id: Long, index: Int, videoTime: Long): Long

  /**
   * 根据时间获取片段下标
   * @param time 传入时间
   * @return 查找到的坐标, 如果查找不到返回-1
   */
  override fun getClipIndex(time: Long): Int {
    return getClipIndex(mId, time)
  }

  private external fun getClipIndex(id: Long, time: Long): Int

  private external fun addAction(): Int

  /**
   * 添加滤镜
   * @param lut 色阶图buffer
   * @param startTime 从哪里开始加
   * @param endTime 到哪里结束
   * @return 滤镜对应的id, 删除或者更新滤镜时需要用到
   */
  override fun addFilter(configPath: String): Int {
    mFilterActionId = addFilter(mId, configPath)
    return mFilterActionId
  }

  private external fun addFilter(id: Long, config: String): Int

  override fun updateFilter(configPath: String, startTime: Int, endTime: Int, actionId: Int) {
    updateFilter(mId, configPath, startTime, endTime, actionId)
  }

  private external fun updateFilter(id: Long, config: String, startTime: Int, endTime: Int, actionId: Int)

  override fun deleteFilter(actionId: Int) {
    deleteFilter(mId, actionId)
  }

  private external fun deleteFilter(id: Long, actionId: Int)

  /**
   * 添加背景音乐
   * @param path 音乐路径
   * @param startTime 从哪里开始
   * @param endTime 到哪里结束
   * @return
   */
  override fun addMusic(config: String): Int {
//    val file = File(path)
//    if (!file.exists() || file.length() == 0L) {
//      // 文件不存在
//      return ErrorCode.FILE_NOT_FOUND
//    }
    return addMusic(mId, config)
  }

  private external fun addMusic(id: Long, config: String): Int

  override fun updateMusic(config: String, actionId: Int) {
    if (mId <= 0) {
      return
    }
    updateMusic(mId, config, actionId)
  }

  private external fun updateMusic(id: Long, config: String, actionId: Int)

  override fun deleteMusic(actionId: Int) {
    if (mId <= 0) {
      return
    }
    deleteMusic(mId, actionId)
  }

  private external fun deleteMusic(id: Long, actionId: Int)

  override fun addAction(configPath: String): Int {
    if (mId <= 0) {
      return -1
    }
    return addAction(mId, configPath)
  }

  private external fun addAction(handle: Long, config: String): Int

  override fun updateAction(startTime: Int, endTime: Int, actionId: Int) {
    if (mId <= 0) {
      return
    }
    updateAction(mId, startTime, endTime, actionId)
  }

  private external fun updateAction(handle: Long, startTime: Int, endTime: Int, actionId: Int)

  override fun deleteAction(actionId: Int) {
    if (mId <= 0) {
      return
    }
    deleteAction(mId, actionId)
  }

  private external fun deleteAction(handle: Long, actionId: Int)

  override fun setOnRenderListener(l: OnRenderListener) {
    mOnRenderListener = l
  }

  /**
   * texture回调
   * 可以做其它特效处理, textureId为普通 TEXTURE_2D 类型
   * @param textureId Int
   * @param width Int
   * @param height Int
   */
  @Suppress("unused")
  private fun onDrawFrame(textureId: Int, width: Int, height: Int): Int {
    return mOnRenderListener?.onDrawFrame(textureId, width, height, null) ?: textureId
  }

  override fun seek(time: Int) {
    if (mId <= 0) {
      return
    }
    seek(mId, time)
  }

  private external fun seek(id: Long, time: Int)

  /**
   * 开始播放
   * @param repeat 是否循环播放
   * @return 播放成功返回0
   */
  override fun play(repeat: Boolean): Int {
    return play(mId, repeat)
  }

  private external fun play(id: Long, repeat: Boolean): Int

  /**
   * 暂停播放
   */
  override fun pause() {
    pause(mId)
  }

  private external fun pause(id: Long)

  /**
   * 继续播放
   */
  override fun resume() {
    resume(mId)
  }

  private external fun resume(id: Long)

  /**
   * 停止播放, 释放资源
   */
  override fun stop() {
    stop(mId)
  }

  private external fun stop(id: Long)

  override fun destroy() {
    release(mId)
    mId = 0
  }

  private external fun release(id: Long)

  @Suppress("unused")
  private fun onPlayStatusChanged(status: Int) {

  }

  @Suppress("unused")
  private fun onPlayError(error: Int) {

  }
}

File: library/src/main/cpp/editor/video_editor.cc
/*
 * Copyright (C) 2019 Trinity. All rights reserved.
 * Copyright (C) 2019 Wang LianJie <wlanjie888@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//
// Created by wlanjie on 2019-05-14.
//

#include "video_editor.h"
#include "android_xlog.h"
#include "gl.h"

namespace trinity {

VideoEditor::VideoEditor(JNIEnv* env, jobject object, const char* resource_path) : Handler()
    , window_(nullptr)
    , video_editor_object_()
    , repeat_(false)
    , play_index_(0)
    , editor_resource_()
    , vm_()
    , queue_mutex_()
    , queue_cond_() {
    window_ = nullptr;
    repeat_ = false;
    player_ = new Player(env, object);
    player_->AddObserver(this);
    editor_resource_ = new EditorResource(resource_path);
    env->GetJavaVM(&vm_);
    video_editor_object_ = env->NewGlobalRef(object);
}

VideoEditor::~VideoEditor() {
    if (nullptr != player_) {
        player_->Destroy();
        delete player_;
        player_ = nullptr;
    }
    if (nullptr != editor_resource_) {
        delete editor_resource_;
        editor_resource_ = nullptr;
    }
    JNIEnv* env = nullptr;
    if ((vm_)->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) == JNI_OK) {
        env->DeleteGlobalRef(video_editor_object_);
    }
}

int VideoEditor::Init() {
    int result = pthread_mutex_init(&queue_mutex_, nullptr);
    if (result != 0) {
        LOGE("init clip queue mutex error");
        return result;
    }
    result = pthread_cond_init(&queue_cond_, nullptr);
    if (result != 0) {
        LOGE("init clip queue cond error");
        return result;
    }
    return result;
}

void VideoEditor::OnSurfaceCreated(jobject surface) {
    if (nullptr != window_) {
        ANativeWindow_release(window_);
        window_ = nullptr;
    }
    JNIEnv* env = nullptr;
    if ((vm_)->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
        return;
    }
    window_ = ANativeWindow_fromSurface(env, surface);
    if (nullptr != player_) {
        player_->OnSurfaceCreated(window_);
    }
}

void VideoEditor::OnSurfaceChanged(int width, int height) {
    if (nullptr != player_) {
        player_->OnSurfaceChanged(width, height);
    }
}

void VideoEditor::OnSurfaceDestroyed() {
    if (nullptr != player_) {
        player_->OnSurfaceDestroy();
    }
    if (nullptr != window_) {
        ANativeWindow_release(window_);
        window_ = nullptr;
    }
}

int64_t VideoEditor::GetVideoDuration() const {
    int64_t duration = 0;
    for (int i = 0; i < clip_deque_.size(); ++i) {
        MediaClip* clip = clip_deque_.at(i);
        duration += clip->end_time - clip->start_time;
    }
    return duration;
}

int64_t VideoEditor::GetCurrentPosition() const {
    if (nullptr != player_) {
        int duration = 0;
        for (int i = 0; i < clip_deque_.size(); i++) {
            if (i < play_index_) {
                MediaClip* clip = clip_deque_.at(i);
                duration += clip->end_time - clip->start_time;
            }
        }
        return duration + player_->GetCurrentPosition();
    }
    return 0;
}

int VideoEditor::GetClipsCount() {
    pthread_mutex_lock(&queue_mutex_);
    int size = static_cast<int>(clip_deque_.size());
    pthread_mutex_unlock(&queue_mutex_);
    return size;
}

MediaClip *VideoEditor::GetClip(int index) {
    if (index < 0 || index >= clip_deque_.size()) {
        return nullptr;
    }
    return clip_deque_.at(index);
}

int VideoEditor::CheckFileType(MediaClip *clip) {
    auto path = clip->file_name;
    FILE* file = fopen(path, "r");
    if (file == nullptr) {
        return -1;
    }
    auto buffer = new uint8_t[12];
    fread(buffer, sizeof(uint8_t), sizeof(uint8_t) * 12, file);
    fclose(file);
    if ((buffer[6] == 'J' && buffer[7] == 'F' && buffer[8] == 'I' && buffer[9] == 'F') ||
        (buffer[6] == 'E' && buffer[7] == 'x' && buffer[8] == 'i' && buffer[9] == 'f')) {
        // JPEG
        clip->type = IMAGE;
    } else if (buffer[1] == 'P' && buffer[2] == 'N' && buffer[3] == 'G') {
        // PNG
        clip->type = IMAGE;
    } else if (buffer[4] == 'f' && buffer[5] == 't' && buffer[6] == 'y' && buffer[7] == 'p') {
        // mp4
        clip->type = VIDEO;
    } else {
        delete[] buffer;
        return -2;
    }
    delete[] buffer;
    return 0;
}

int VideoEditor::InsertClip(MediaClip *clip) {
    int ret = CheckFileType(clip);
    if (ret != 0) {
        return ret;
    }
    pthread_mutex_lock(&queue_mutex_);
    clip_deque_.push_back(clip);
    editor_resource_->InsertClip(clip);
    pthread_mutex_unlock(&queue_mutex_);
    return 0;
}

int VideoEditor::InsertClipWithIndex(int index, MediaClip *clip) {
    return 0;
}

void VideoEditor::RemoveClip(int index) {
    if (index < 0 || index > clip_deque_.size()) {
        return;
    }
    editor_resource_->RemoveClip(index);
    MediaClip* clip = clip_deque_.at(index);
    delete clip;
    clip_deque_.erase(clip_deque_.begin() + index);
}

int VideoEditor::ReplaceClip(int index, MediaClip *clip) {
    editor_resource_->ReplaceClip(index, clip);
    return 0;
}

int64_t VideoEditor::GetVideoTime(int index, int64_t clip_time) {
    if (index < 0 || index >= clip_deque_.size()) {
        return 0;
    }
    int64_t offset = 0;
    for (int i = 0; i < index; ++i) {
        MediaClip* clip = clip_deque_.at(i);
        offset += (clip->end_time - clip->start_time);
    }
    return clip_time + offset;
}

int64_t VideoEditor::GetClipTime(int index, int64_t video_time) {
    if (index < 0 || index >= clip_deque_.size()) {
        return 0;
    }
    int64_t offset = 0;
    for (int i = 0; i < index; ++i) {
        MediaClip* clip = clip_deque_.at(i);
        offset += (clip->end_time - clip->start_time);
    }
    return video_time - offset;
}

int VideoEditor::GetClipIndex(int64_t time) {
    return 0;
}

int VideoEditor::AddFilter(const char *config) {
    if (nullptr != player_) {
        int action_id = player_->AddFilter(config);
        if (nullptr != editor_resource_) {
            editor_resource_->AddFilter(config, action_id);
        }
        return action_id;
    }
    return -1;
}

void VideoEditor::UpdateFilter(const char *config, int start_time, int end_time, int action_id) {
    if (nullptr != player_) {
        player_->UpdateFilter(config, start_time, end_time, action_id);
        if (nullptr != editor_resource_) {
            editor_resource_->UpdateFilter(config, start_time, end_time, action_id);
        }
    }
}

void VideoEditor::DeleteFilter(int action_id) {
    if (nullptr != player_) {
        player_->DeleteFilter(action_id);
        if (nullptr != editor_resource_) {
            editor_resource_->DeleteFilter(action_id);
        }
    }
}

int VideoEditor::AddMusic(const char* music_config) {
    if (nullptr != player_) {
        int action_id = player_->AddMusic(music_config);
        if (nullptr != editor_resource_) {
            editor_resource_->AddMusic(music_config, action_id);
        }
        return action_id;
    }
    return -1;
}

void VideoEditor::UpdateMusic(const char* music_config, int action_id) {
    if (nullptr != player_) {
        player_->UpdateMusic(music_config, action_id);

        if (nullptr != editor_resource_) {
            editor_resource_->UpdateMusic(music_config, action_id);
        }
    }
}

void VideoEditor::DeleteMusic(int action_id) {
    if (nullptr != player_) {
        player_->DeleteMusic(action_id);
        if (nullptr != editor_resource_) {
            editor_resource_->DeleteMusic(action_id);
        }
    }
}

int VideoEditor::AddAction(const char *effect_config) {
    if (nullptr != player_) {
        auto action_id = player_->AddAction(effect_config);
        if (nullptr != editor_resource_) {
            editor_resource_->AddAction(effect_config, action_id);
        }
        return action_id;
    }
    return -1;
}

void VideoEditor::UpdateAction(int start_time, int end_time, int action_id) {
    if (nullptr != player_) {
        player_->UpdateAction(start_time, end_time, action_id);
        if (nullptr != editor_resource_) {
            editor_resource_->UpdateAction(start_time, end_time, action_id);
        }
    }
}

void VideoEditor::DeleteAction(int action_id) {
    if (nullptr != player_) {
        player_->DeleteAction(action_id);
    }
    if (nullptr != editor_resource_) {
        editor_resource_->DeleteAction(action_id);
    }
}

void VideoEditor::OnComplete() {
    LOGE("enter %s", __func__);
    if (clip_deque_.size() == 1) {
        play_index_ = repeat_ ? 0 : -1;
    } else {
        play_index_++;
        if (play_index_ >= clip_deque_.size()) {
            play_index_ = repeat_ ? 0 : -1;
        }
    }
    if (play_index_ == -1) {
        return;
    }

    int video_count_duration = 0;
    for (int i = 0; i < clip_deque_.size(); i++) {
        if (i < play_index_) {
            auto* clip = clip_deque_.at(static_cast<unsigned int>(i));
            video_count_duration += (clip->end_time - clip->start_time);
        }
    }

    MediaClip* clip = clip_deque_.at(static_cast<unsigned int>(play_index_));
    if (nullptr != player_) {
        player_->Start(clip, video_count_duration);
    }
    LOGE("leave %s", __func__);
}

void VideoEditor::Seek(int time) {
    if (nullptr != player_) {
        player_->Seek(time, INT_MAX);
    }
}

int VideoEditor::Play(bool repeat, JNIEnv* env, jobject object) {
    if (clip_deque_.empty()) {
        return -1;
    }
    repeat_ = repeat;
    MediaClip* clip = clip_deque_.at(0);

    if (nullptr != player_) {
        return player_->Start(clip);
    }
    return 0;
}

void VideoEditor::Pause() {
    if (nullptr != player_) {
        player_->Pause();
    }
}

void VideoEditor::Resume() {
    if (nullptr != player_) {
        player_->Resume();
    }
}

void VideoEditor::Stop() {}

void VideoEditor::Destroy() {
    LOGI("enter Destroy");
    if (nullptr != player_) {
        player_->Destroy();
    }

    pthread_mutex_lock(&queue_mutex_);
    for (auto clip : clip_deque_) {
        delete[] clip->file_name;
        delete clip;
    }
    clip_deque_.clear();
    pthread_mutex_unlock(&queue_mutex_);
    pthread_mutex_destroy(&queue_mutex_);
    pthread_cond_destroy(&queue_cond_);
    LOGI("leave Destroy");
}

}  // namespace trinity
